protocol/0053-PERP-product_builtin_perpetual_future.md

```diff
@@ -53,33 +56,102 @@ cash_settled_perpetual_future.value(quote) {
+
+- a field on the oracle payload,
+- a timestamp from the oracle's blockchain,
+- an internal vega time.
 
 ### 4.2. Mark to market settlement
 
-Every time a [mark to market settlement](./0003-MTMK-mark_to_market_settlement.md) is carried out the value of the last periodic settlement data point received along with the price used for MTM settlement and the current `vegaprotocol.builtin.timestamp` gets stored as the funding payment data point. If no periodic settlement data has been received yet then the funding payment data point should not be created.
+Every time a [mark to market settlement](./0003-MTMK-mark_to_market_settlement.md) is carried out the value of mark price (`f`) and the current `vegaprotocol.builtin.timestamp` gets stored as an internal data point within the market.
 
 ### 4.3. Periodic settlement
 
-When the `settlement_schedule` event is received the latest funding payment data point gets repeated with the timestamp set to the current value of `vegaprotocol.builtin.timestamp`.
+When the `settlement_schedule` event is received we need to calculated the funding payment. Store the current vega time as `funding_period_end`.
+
+If there are no oracle data points with a timestamp less than `funding_period_end` available then funding payment is skipped and `funding_period_start` gets overwritten with `funding_period_end`.
+
+If such points available then the calculations discussed in the following subsections get executed and funding payments get exchanged.
+
+#### TWAP spot price calculation
+
+Traverse all the available oracle data point tuples `(s,t)` and calculate the time-weighted average spot price (`s_twap`) as:
+
+```go
+var previous_point
+sum_product := 0
+
+for p := range oracle_data_points {
+    if p.t <= funding_period_start {
+        previous_point = p
+        continue
+    }
+    if p.t >= funding_period_end {
+        break
+    }
+    if previous_point != nil {
+        sum_product += previous_point*(p.t-max(funding_period_start,previous_point.t))
+    }
+    previous_point = p
+}
+
+sum_product += previous_point.s*(funding_period_end-max(funding_period_start,previous_point.t))
+s_twap = sum_product / (funding_period_end - max(funding_period_start, oracle_data_points[0].t))
+```
 
-The next step is to calculate the periodic settlement funding payment. If there are no periodic settlement data points then the periodic settlement is skipped. Otherwise, consider all the periodic settlement data points and calculate the time-weighted average price difference as:
+Only the oracle data point with largest timestamp that's less than or equal to `funding_period_end` (and any data points with larger timestamps) need to be kept from that point on.
```

Comment:
An edge case that has been sitting in my head for a few days:
Given that the timestamp we are associating with the oracle data point may come from a different chain _is it at all possible_ that we might have an oracle data-point that comes in which _for whatever reason_ is a little bit in the future from vega's point-of-view of "now". Would we want want keep those for when they do become within the funding period?

Also another thing:
When we trigger a the periodic settlement from the incoming oracle data, I'm assuming we'll do it at the end of the block? I'm thinking we would also want to processing any oracle data that might also be waiting in to be processed in the same block but just happened to be after the oracle tx was processed. Also MTM happens at the end of a block and we'd want to include that point.

Reply:
I guess if the external timestamp says it's ahead of vega time (ahead of a funding payment) then simply don't include data associated to that timepoint in the funding calculation but it will play a role in the next one.